solidity/contracts/relay/Relay.sol

```diff
@@ -0,0 +1,507 @@
+    /// same difficulty; headers from the latter epoch will be accepted if the
+    /// previous epoch has been proven to the relay.
+    /// This is because it is not possible to distinguish such headers from
+    /// headers of the previous epoch.
+    ///
+    /// If the difficulty increases significantly between relay genesis and the
+    /// present, creating fraudulent proofs for earlier epochs becomes easier.
+    /// Users of the relay should check the timestamps of valid headers and
+    /// only accept appropriately recent ones.
+    function validateChain(bytes memory headers)
+        external
+        view
+        relayActive
+        returns (uint256)
+    {
+        require(headers.length % 80 == 0, "Invalid header length");
+
+        uint256 headerCount = headers.length / 80;
+
+        require(
+            headerCount > 1 && headerCount < 2016,
+            "Invalid number of headers"
+        );
+
+        uint256 currentHeaderTimestamp = headers.extractTimestamp();
+
+        uint256 relevantEpoch = currentEpoch;
+
+        // timestamp of the epoch the header chain starts in
+        uint256 startingEpochTimestamp = epochs[currentEpoch].timestamp;
+        // timestamp of the next epoch after that
+        uint256 nextEpochTimestamp = 0;
+
+        // Find the correct epoch for the given chain
+        // Fastest with recent epochs, but able to handle anything after genesis
+        while (currentHeaderTimestamp < startingEpochTimestamp) {
+            relevantEpoch -= 1;
+            nextEpochTimestamp = startingEpochTimestamp;
+            startingEpochTimestamp = epochs[relevantEpoch].timestamp;
+            require(
+                // This works because every epoch starting from genesis has
+                // a recorded timestamp, so a timestamp of zero means we have
+                // reached before the genesis.
+                startingEpochTimestamp > 0,
+                "Cannot validate chains before relay genesis"
+            );
+        }
+
+        uint256 relevantTarget = epochs[relevantEpoch].target;
+
+        // Short-circuit the first header's validation.
+        (
+            bytes32 previousHeaderDigest,
+            uint256 currentHeaderTarget
+        ) = validateHeader(headers, 0, bytes32(0));
+
+        require(
+            currentHeaderTarget == relevantTarget,
+            "Invalid target in header chain"
+        );
+
+        for (uint256 i = 1; i < headerCount; i++) {
+            uint256 previousHeaderTimestamp = currentHeaderTimestamp;
+            bytes32 currentDigest;
+            (currentDigest, currentHeaderTarget) = validateHeader(
+                headers,
+                i * 80,
+                previousHeaderDigest
+            );
+
+            currentHeaderTimestamp = headers.extractTimestampAt(i * 80);
+
+            require(
+                currentHeaderTimestamp > previousHeaderTimestamp,
+                "Invalid timestamp in header chain"
+            );
```

Comment:
I am not sure if we should validate it this way. 

> A timestamp is accepted as valid if it is greater than the median timestamp of previous 11 blocks, and less than the network-adjusted time + 2 hours. "Network-adjusted time" is the median of the timestamps returned by all nodes connected to you. As a result block timestamps are not exactly accurate, and they do not need to be. Block times are accurate only to within an hour or two.

https://en.bitcoin.it/wiki/Block_timestamp 

If headers form a valid chain (we validate the previous header digest), and we checked the difficulty target for all of them (it's done in `validateHeader`), do we still need to check timestamps?